### PR TITLE
added subgraph

### PR DESCRIPTION
**Pull request by Clearwood**
Please remember the [Contributing Guidelines](https://WhiteRabbitTheWorldIsYours/dashboard/blob/master/CONTRIBUTION.md) :heart:

## What does this PR do?

Creates a subgraph indexing Deposits

## Recommendations for testing

Playground: https://thegraph.com/explorer/subgraph/clearwood/white-rabbit-cp
# GraphQL based API
Query URL: `https://api.thegraph.com/subgraphs/name/clearwood/white-rabbit-cp`
# Available requests
## Get Movie by ID
```
{
  movie(id: 1343237633) {
    id
    total
  }
}
```

Example response:

```
{
  "data": {
    "movie": {
      "id": "1343237633",
      "total": "210000000000000000"
    }
  }
}
```

## Links to relevant issues or information

Closes https://github.com/WhiteRabbitTheWorldIsYours/dashboard/issues/57

### CODE REVIEW

File: subgraph/package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "dashboard",
+  "license": "UNLICENSED",
```
Comment (troggy) [inline subgraph/package.json:3]: proper license here?

File: subgraph/package.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "dashboard",
```
Comment (troggy) [inline subgraph/package.json:2]:
```suggestion
  "name": "@whiterabbitjs/dashboard-subgraph",
```

File: subgraph/package.json (hunk 1 above)
Comment (Clearwood) [reply subgraph/package.json:3]: what License would that be?

Comment (troggy) [reply subgraph/package.json:3]: same as in other repos?

**Comment by troggy** [comment]
how do I test it? Any suggestions?

**Comment by Clearwood** [comment]
> how do I test it? Any suggestions?

here you could use the playground: https://thegraph.com/explorer/subgraph/clearwood/white-rabbit-cp

File: subgraph/schema.graphql [hunk 3]
```diff
@@ -0,0 +1,12 @@
+type Region @entity {
```
Comment (troggy) [inline subgraph/schema.graphql:1]: the name is a bit confusing like it is a total per region, while it is total per movie for the region

Comment (Clearwood) [reply subgraph/schema.graphql:1]: it is a total per region per movie.

Comment (Clearwood) [reply subgraph/schema.graphql:1]: the idea was for it to be a type of region, so I could map the one to many relationship of movie to region

Comment (Clearwood) [reply subgraph/schema.graphql:1]: would you like better commenting or a different name? if so do you have an idea what would make more sense

Comment (troggy) [reply subgraph/schema.graphql:1]: maybe call it `RevenuePerMovieRegion` and the other one `RevenuePerMovie` ? to be precise

File: subgraph/subgraph.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+specVersion: 0.0.2
+schema:
+  file: ./schema.graphql
+dataSources:
+  - kind: ethereum/contract
+    name: Contract
```
Comment (troggy) [inline subgraph/subgraph.yaml:6]: can we name it CollectionPool for clarity? And ABI file

File: subgraph/schema.graphql (hunk 3 above)
Comment (troggy) [reply subgraph/schema.graphql:1]: I'd use the same common sense as with any other entities: ideally the name should tell what is the entity represents. If it represents the movie it should be ideally named `Movie`. If it represents a revenue for the movie, then named something like `MovieRevenue` or `RevenuePerMovie`. It is just easier to read the code with clearly named entities, sparing some braincycles for humans later on

Comment (troggy) [reply subgraph/schema.graphql:1]: you obviously cannot add comments everywhere you use the name, so changing the name is easier + more DRY

File: subgraph/package.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "@whiterabbitjs/dashboard-subgraph",
+  "license": "UNLICENSED",
+  "scripts": {
+    "codegen": "graph codegen",
+    "build": "graph build",
+    "deploy": "graph deploy --node https://api.thegraph.com/deploy/ --ipfs https://api.thegraph.com/ipfs/ clearwood/white-rabbit-cp",
+    "create-local": "graph create --node http://localhost:8020/ clearwood/white-rabbit-cp",
```
Comment (troggy) [inline subgraph/package.json:8]: do we need these scripts? If yes can we have some description in README about them please?

Comment (troggy) [reply subgraph/package.json:8]: I guess they require local Graph Node to be started?

File: subgraph/package.json (hunk 1 above)
Comment (Clearwood) [reply subgraph/package.json:3]: So far no WR repo has a license

File: subgraph/schema.graphql [hunk 6]
```diff
@@ -1,12 +1,12 @@
-type Region @entity {
+type RevenuePerMovieRegion @entity {
   id: ID!
-  movie: Movie!
+  revenuePerMovie: RevenuePerMovie!
```
Comment (troggy) [inline subgraph/schema.graphql:3]: shouldn't it be `movieId: string` here as per https://github.com/WhiteRabbitTheWorldIsYours/dashboard/pull/58/commits/715ac3b6be2f7b1fcb18a022db8728b3f3decb59#diff-4a7df0b7ce9a0a0e9606c9afb21c5774R27 ?

Comment (troggy) [reply subgraph/schema.graphql:3]: I mean, line 27 assigns `movieId` here, so both the name and type are incorrect, no?

**Comment by troggy** [comment]
do I need you API KEY to update https://api.thegraph.com/subgraphs/name/clearwood/white-rabbit-cp/graphql endpoint? Can we have a shared one (with API KEY shared privately within WR devs) ?

File: subgraph/schema.graphql (hunk 6 above)
Comment (troggy) [reply subgraph/schema.graphql:3]: image

Comment (troggy) [inline subgraph/schema.graphql:3]:
```suggestion
  movieId: String!
```

File: subgraph/schema.graphql [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
   region: Int!
   total: BigInt!
 }
 
-type Movie @entity {
+type RevenuePerMovie @entity {
   id: ID!
   total: BigInt!
-  regions: [Region!]! @derivedFrom(field: "movie")
+  revenuePerMovieRegions: [RevenuePerMovieRegion!]! @derivedFrom(field: "revenuePerMovie")
```
Comment (troggy) [inline subgraph/schema.graphql:11]:
```suggestion
  perRegion: [RevenuePerMovieRegion!]! @derivedFrom(field: "revenuePerMovie")
```

File: subgraph/src/mapping.ts [hunk 8]
```diff
@@ -5,26 +5,26 @@ import {
...
+    movie = new RevenuePerMovie(movieId)
     movie.total = BigInt.fromI32(0)
   }
   if (region == null) {
-    region = new Region(regionId)
+    region = new RevenuePerMovieRegion(regionId)
     region.total = BigInt.fromI32(0)
     region.region = event.params.region
-    region.movie = movieId
+    region.revenuePerMovie = movieId
```
Comment (troggy) [inline subgraph/src/mapping.ts:27]:
```suggestion
    region.movieId = movieId
```

File: subgraph/schema.graphql (hunk 6 above)
Comment (troggy) [reply subgraph/schema.graphql:3]: ah, I see it needed for one-to-many relationship 😅

**Comment by Clearwood** [comment]
> this works . Approved. Love that so little code needed 
> 
> Not sure how to redeploy shared thegraph endpoint once it is used in the other modules (probably API key needs to be shared)

probably best to redeploy it as the github organization but I had problems requesting access for that, otherwise I'd have to share the auth token.

**Pull request merged**